File: src/test/java/cube8540/oauth/authentication/users/domain/UserTest.java
package cube8540.oauth.authentication.users.domain;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.crypto.password.PasswordEncoder;

import static cube8540.oauth.authentication.users.domain.UserTestHelper.APPROVAL_AUTHORITY;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.AUTHORITY;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.CHANGE_PASSWORD;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.CLIENT_ID;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.ENCRYPTED_PASSWORD;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.PASSWORD;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.RAW_CREDENTIALS_KEY;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.RAW_PASSWORD_CREDENTIALS_KEY;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.UID;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeCredentialsKey;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeExpiredKey;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeKeyGenerator;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeMismatchedKey;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeMismatchesEncoder;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makePassValidatorFactory;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makePasswordEncoder;
import static cube8540.oauth.authentication.users.domain.UserTestHelper.makeUidGenerator;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DisplayName("유저 계정 테스트")
class UserTest {

    @Test
    @DisplayName("유효 하지 않은 정보가 저장 되어 있을시")
    void userDataInvalid() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserValidatorFactory factory = makeErrorValidatorFactory(user);

        assertThrows(UserInvalidException.class, () -> user.validation(factory));
    }

    @Test
    @DisplayName("모든 데이터가 유효할시")
    void userDataAllowed() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserValidatorFactory factory = makePassValidatorFactory(user);

        assertDoesNotThrow(() -> user.validation(factory));
    }

    @Test
    @DisplayName("패스워드 변경시 이전에 사용 하던 패스워드와 일치 하지 않을시")
    void changePasswordWhenNotMatchedExistingPassword() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);
        PasswordEncoder encoder = makeMismatchesEncoder(PASSWORD, ENCRYPTED_PASSWORD);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.changePassword(PASSWORD, CHANGE_PASSWORD, encoder));
        Assertions.assertEquals(UserErrorCodes.INVALID_PASSWORD, e.getCode());
    }

    @Test
    @DisplayName("패스워드 변경")
    void changePassword() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCRYPTED_PASSWORD);

        user.encrypted(encoder);

        user.changePassword(PASSWORD, CHANGE_PASSWORD, encoder);
        assertEquals(CHANGE_PASSWORD, user.getPassword());
    }

    @Test
    @DisplayName("패스워드 분실")
    void forgotPassword() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);
        UserCredentialsKey key = makeCredentialsKey(RAW_PASSWORD_CREDENTIALS_KEY);
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);

        user.forgotPassword(generator);

        assertEquals(key, user.getPasswordCredentialsKey());
    }

    @Test
    @DisplayName("패스워드 인증키가 할당 되지 않은 계정의 패스워드 초기화")
    void resetPasswordToNotGeneratedCredentialsKeyUser() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.resetPassword(RAW_PASSWORD_CREDENTIALS_KEY, CHANGE_PASSWORD));
        assertEquals(UserErrorCodes.INVALID_KEY, e.getCode());
    }

    @Test
    @DisplayName("패스워드 인증키가 매칭 되지 않은 계정의 패스워드 초기화")
    void resetPasswordKeyIsNotMatched() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKey key = makeMismatchedKey();
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);
        user.forgotPassword(generator);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.resetPassword(RAW_PASSWORD_CREDENTIALS_KEY, CHANGE_PASSWORD));
        assertEquals(UserErrorCodes.INVALID_KEY, e.getCode());
    }

    @Test
    @DisplayName("패스워드 인증키가 만료된 계정의 패스워드 초기화")
    void resetPasswordKeyIsExpired() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKey key = makeExpiredKey();
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);
        user.forgotPassword(generator);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.resetPassword(RAW_PASSWORD_CREDENTIALS_KEY, CHANGE_PASSWORD));
        assertEquals(UserErrorCodes.KEY_EXPIRED, e.getCode());
    }

    @Test
    @DisplayName("패스워드 초기화")
    void resetPassword() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKey key = makeCredentialsKey(RAW_CREDENTIALS_KEY);
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);
        user.forgotPassword(generator);

        user.resetPassword(RAW_CREDENTIALS_KEY, CHANGE_PASSWORD);
        assertEquals(CHANGE_PASSWORD, user.getPassword());
        assertNull(user.getPasswordCredentialsKey());
    }

    @Test
    @DisplayName("인증키 할당")
    void grantedCredentialsKey() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKey key = makeCredentialsKey(RAW_CREDENTIALS_KEY);
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);

        user.generateCredentialsKey(generator);
        assertEquals(key, user.getCredentialsKey());
    }

    @Test
    @DisplayName("이미 인증 받은 계정에 인증키 할당")
    void grantedCredentialsKeyToAlreadyCredentialsUser() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKey key = makeCredentialsKey(RAW_CREDENTIALS_KEY);
        UserCredentialsKeyGenerator generator = makeKeyGenerator(key);
        user.generateCredentialsKey(generator);
        user.credentials(RAW_CREDENTIALS_KEY);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.generateCredentialsKey(generator));
        assertEquals(UserErrorCodes.ALREADY_CREDENTIALS, e.getCode());
    }

    @Test
    @DisplayName("인증키가 할당 되지 않은 유저 계정 인증")
    void accountAuthenticationToNotGeneratedKeyUser() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.credentials(RAW_CREDENTIALS_KEY));
        assertEquals(UserErrorCodes.INVALID_KEY, e.getCode());
    }

    @Test
    @DisplayName("인증키가 매칭 되지 않은 계정 인증")
    void accountAuthenticationKeyIsNotMatched() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKeyGenerator generator = makeKeyGenerator(makeMismatchedKey());
        user.generateCredentialsKey(generator);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.credentials(RAW_CREDENTIALS_KEY));
        assertEquals(UserErrorCodes.INVALID_KEY, e.getCode());
    }

    @Test
    @DisplayName("인증키가 만료된 계정 인증")
    void accountAuthenticationKeyIsExpired() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKeyGenerator generator = makeKeyGenerator(makeExpiredKey());
        user.generateCredentialsKey(generator);

        UserAuthorizationException e = assertThrows(UserAuthorizationException.class, () -> user.credentials(RAW_CREDENTIALS_KEY));
        assertEquals(UserErrorCodes.KEY_EXPIRED, e.getCode());
    }

    @Test
    @DisplayName("계정 인증")
    void accountAuthentication() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        UserCredentialsKeyGenerator generator = makeKeyGenerator(makeCredentialsKey(RAW_CREDENTIALS_KEY));
        user.generateCredentialsKey(generator);

        user.credentials(RAW_CREDENTIALS_KEY);
        assertNull(user.getCredentialsKey());
        assertTrue(user.getCredentialed());
    }

    @Test
    @DisplayName("패스워드 암호화")
    void passwordEncrypting() {
        PasswordEncoder encoder = makePasswordEncoder(PASSWORD, ENCRYPTED_PASSWORD);
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        user.encrypted(encoder);
        assertEquals(ENCRYPTED_PASSWORD, user.getPassword());
    }

    @Test
    @DisplayName("유저 승인 권한 추가")
    void addApprovalAuthority() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        user.addApprovalAuthority(CLIENT_ID, AUTHORITY);
        assertTrue(user.getApprovalAuthorities().contains(APPROVAL_AUTHORITY));
    }

    @Test
    @DisplayName("유저 승인 권한이 null 일때 승인 권한 삭제")
    void revokeApprovalAuthorityWhenUserApprovalAuthoritiesIsNull() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        user.setApprovalAuthorities(null);

        assertDoesNotThrow(() -> user.revokeApprovalAuthority(CLIENT_ID, AUTHORITY));
    }

    @Test
    @DisplayName("유저 승인 권한 삭제")
    void revokeApprovalAuthority() {
        User user = new User(makeUidGenerator(UID), RAW_USERNAME, PASSWORD);

        user.addApprovalAuthority(CLIENT_ID, AUTHORITY);
        user.revokeApprovalAuthority(CLIENT_ID, AUTHORITY);

        assertFalse(user.getApprovalAuthorities().contains(APPROVAL_AUTHORITY));
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/user/domain/UserTest.kt
package cube8540.oauth.authentication.user.domain

import cube8540.oauth.authentication.users.domain.*
import cube8540.validator.core.ValidationError
import cube8540.validator.core.ValidationRule
import cube8540.validator.core.Validator
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.springframework.security.crypto.password.PasswordEncoder

class UserTest {

    @Nested
    inner class ValidationTest {
        private val userValidatorFactory = mockk<UserValidatorFactory>()
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator>()
            every { uidGenerator.generateUid() } returns Uid("UID")

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `user data is invalid`() {
            val rule = mockk<ValidationRule<User>> {
                every { isValid(user) } returns false
                every { error() } returns ValidationError("username", "test")
            }
            every { userValidatorFactory.createValidator(user) } returns Validator.of(user).registerRule(rule)

            val thrown = catchThrowable { user.validation(userValidatorFactory) }
            assertThat(thrown).isInstanceOf(UserInvalidException::class.java)
        }

        @Test
        fun `user data is allowed`() {
            val rule = mockk<ValidationRule<User>> {
                every { isValid(user) } returns true
            }
            every { userValidatorFactory.createValidator(user) } returns Validator.of(user).registerRule(rule)

            assertThatCode { user.validation(userValidatorFactory) }.doesNotThrowAnyException()
        }
    }

    @Nested
    inner class EncryptingTest {
        private val user: User
        private val encoder = mockk<PasswordEncoder>()

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `password encrypting`() {
            every { encoder.encode("password") } returns "encodingPassword"

            user.encrypted(encoder)
            assertThat(user.password).isEqualTo("encodingPassword")
        }
    }

    @Nested
    inner class PasswordChangeTest {
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `existing password is not equal`() {
            val encoder = mockk<PasswordEncoder> {
                every { matches("existingPassword", "password") } returns false
            }

            val thrown = catchThrowable { user.changePassword("existingPassword", "newPassword", encoder) }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.INVALID_PASSWORD)
        }

        @Test
        fun `change password successful`() {
            val encoder = mockk<PasswordEncoder> {
                every { matches("existingPassword", "password") } returns true
            }

            user.changePassword("existingPassword", "newPassword", encoder)
            assertThat(user.password).isEqualTo("newPassword")
        }
    }

    @Nested
    inner class GenerateCredentialsKeyTest {
        private val credentialsKey = mockk<UserCredentialsKey>()
        private val generator = mockk<UserCredentialsKeyGenerator>()
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }
            every { generator.generateKey() } returns credentialsKey

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `generate credentials key for user`() {
            user.generateCredentialsKey(generator)

            assertThat(user.credentialsKey).isEqualTo(credentialsKey)
        }

        @Test
        fun `user account is already credentials`() {
            accountCredentials()

            val thrown = catchThrowable { user.generateCredentialsKey(generator) }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code)
                .isEqualTo(UserErrorCodes.ALREADY_CREDENTIALS)
        }

        private fun accountCredentials() {
            user.generateCredentialsKey(generator)
            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.MATCHED
            user.credentials("key")
        }
    }

    @Nested
    @TestInstance(TestInstance.Lifecycle.PER_METHOD)
    inner class CredentialsTest {
        private val credentialsKey = mockk<UserCredentialsKey>()
        private val generator = mockk<UserCredentialsKeyGenerator>()
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }
            every { generator.generateKey() } returns credentialsKey

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `credentials key is not generated`() {
            val thrown = catchThrowable { user.credentials("key") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.INVALID_KEY)
        }

        @Test
        fun `credentials key is not matches`() {
            user.generateCredentialsKey(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.NOT_MATCHED

            val thrown = catchThrowable { user.credentials("key") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.INVALID_KEY)
        }

        @Test
        fun `credentials key is expired`() {
            user.generateCredentialsKey(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.EXPIRED

            val thrown = catchThrowable { user.credentials("key") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.KEY_EXPIRED)
        }

        @Test
        fun `credentials is successful`() {
            user.generateCredentialsKey(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.MATCHED

            user.credentials("key")
            assertThat(user.credentialed).isTrue
            assertThat(user.credentialsKey).isNull()
        }
    }

    @Nested
    inner class ForgotPasswordTest {
        private val credentialsKey = mockk<UserCredentialsKey>()
        private val generator = mockk<UserCredentialsKeyGenerator>()
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }
            every { generator.generateKey() } returns credentialsKey

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `request forgot password`() {
            user.forgotPassword(generator)

            assertThat(user.passwordCredentialsKey).isEqualTo(credentialsKey)
        }
    }

    @Nested
    inner class ResetPasswordTest {
        private val credentialsKey = mockk<UserCredentialsKey>()
        private val generator = mockk<UserCredentialsKeyGenerator>()
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }
            every { generator.generateKey() } returns credentialsKey

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `password credentials key is not generated`() {
            val thrown = catchThrowable { user.resetPassword("key", "newPassword") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.INVALID_KEY)
        }

        @Test
        fun `password credentials key is not matches`() {
            user.forgotPassword(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.NOT_MATCHED

            val thrown = catchThrowable { user.resetPassword("key", "newPassword") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.INVALID_KEY)
        }

        @Test
        fun `password credentials key is expired`() {
            user.forgotPassword(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.EXPIRED

            val thrown = catchThrowable { user.resetPassword("key", "newPassword") }
            assertThat(thrown).isInstanceOf(UserAuthorizationException::class.java)
            assertThat((thrown as UserAuthorizationException).code).isEqualTo(UserErrorCodes.KEY_EXPIRED)
        }

        @Test
        fun `reset password successful`() {
            user.forgotPassword(generator)

            every { credentialsKey.matches("key") } returns UserKeyMatchedResult.MATCHED

            user.resetPassword("key", "newPassword")
            assertThat(user.password).isEqualTo("newPassword")
            assertThat(user.passwordCredentialsKey).isNull()
        }
    }

    @Nested
    inner class AddApprovalAuthorityTest {
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `add approval successful`() {
            user.addApprovalAuthority("clientId", "approval")

            val expected = ApprovalAuthority("clientId", "approval")
            assertThat(user.approvalAuthorities).containsExactly(expected)
        }
    }

    @Nested
    inner class RevokeApprovalAuthorityTest {
        private val user: User

        init {
            val uidGenerator = mockk<UserUidGenerator> {
                every { generateUid() } returns Uid("UID")
            }

            this.user = User(uidGenerator, "username", "password")
        }

        @Test
        fun `revoke when approval authorities is null`() {
            assertThatCode { user.revokeApprovalAuthority("clientId", "approval") }
                .doesNotThrowAnyException()
        }

        @Test
        fun `revoke authority is successful`() {
            user.addApprovalAuthority("clientId", "approval")

            user.revokeApprovalAuthority("clientId", "approval")

            val expected = ApprovalAuthority("clientId", "approval")
            assertThat(user.approvalAuthorities).doesNotContain(expected)
        }
    }
}